src/meta/api/src/schema_api_impl.rs

```diff
@@ -2855,3 +2909,83 @@ async fn list_tables_from_share_db(
     )
     .await
 }
+
+async fn build_upsert_table_copied_file_info_conditions<T>(
+    this: &T,
+    req: &UpsertTableCopiedFileReq,
+    file_name_keys: &[String],
+    tb_meta_seq: u64,
+    fail_if_duplicated: bool,
+) -> Result<(Vec<TxnCondition>, Vec<TxnOp>), KVAppError>
+where
+    T: kvapi::KVApi<Error = MetaError>,
+{
+    let table_id = req.table_id;
+    let tbid = TableId { table_id };
+
+    let mut condition = vec![txn_cond_seq(&tbid, Eq, tb_meta_seq)];
+    let mut if_then = vec![];
+
+    // TODO put this doc in upsert_table_copied_file_info?
+    // `remove_table_copied_files` and `upsert_table_copied_file_info`
+    // all modify `TableCopiedFileInfo`,
+    // so there used to has `TableCopiedFileLockKey` in these two functions
+    // to protect TableCopiedFileInfo modification.
+    // In issue: https://github.com/datafuselabs/databend/issues/8897,
+    // there is chance that if copy files concurrently, `upsert_table_copied_file_info`
+    // may return `TxnRetryMaxTimes`.
+    // So now, in case that `TableCopiedFileInfo` has expire time, remove `TableCopiedFileLockKey`
+    // in each function. In this case there is chance that some `TableCopiedFileInfo` may not be
+    // removed in `remove_table_copied_files`, but these data can be purged in case of expire time.
+
+    let mut file_name_infos = req.file_info.clone().into_iter();
+
+    if fail_if_duplicated {
+        // "fail_if_duplicated" mode, assumes files are absent
+        for (file_name, file_info) in file_name_infos {
+            let key = TableCopiedFileNameIdent {
+                table_id,
+                file: file_name.to_owned(),
+            };
+            condition.push(txn_cond_seq(&key, Eq, 0));
```

Comment:
not need `condition.push(txn_cond_seq(&key, Eq, 0));` if  table_verison not changed sinces https://github.com/datafuselabs/databend/blob/main/src/query/service/src/interpreters/interpreter_copy.rs#L276

Reply:
we can also discuss it later without blocking this pr.

Reply:
> but this maybe a conflict if we will optimize force = true later out of txn?

not sure.

> not need condition.push(txn_cond_seq(&key, Eq, 0)); if table_verison not changed since

since we have already checked the files during the "coloring" phase? oh, it is reasonable to me, let me try this

Reply:
oh, it is not safe to do that. 

 since the SchemaApi exposes `upsert_table_copied_file_info`, the table's copied files can be updated without increasing the version of the table.